### PR TITLE
samples: net: irc_bot: simplify and update irc_bot sample

### PR DESCRIPTION
**Pull request by mike-scott**
Originally, irc_bot sample was intended to be a prototype for setting
up an IRC-based parsing / handler library in subsys/net/lib.  For this
purpose it has several structures (irc and channel related) to allow
for individual channel based message handlers.

This effort has faded over time and the sample itself ended up being
very complex for the average user to "play around" with.

Let's simplify this sample by removing the irc/channel structures
and update it to use the full net_app APIs.  Also, fix the parse
issues across multiple net bufs.

These changes remove over 400 lines of code from this sample (~30%
size reduction).

Signed-off-by: Michael Scott <michael.scott@linaro.org>

### CODE REVIEW

**Comment by mike-scott** [comment]
I considered breaking this up into several small patches, but they all seemed to touch each other so I ended up with 1 rather large patch.  :/  Recommendations are welcome.

File: samples/net/irc_bot/src/irc-bot.c [hunk 1]
```diff
@@ -43,179 +45,144 @@ static u8_t cmd_buf[CMD_BUFFER_SIZE];
...
 #endif
-
-/* DNS API */
-#define DNS_QUERY_TIMEOUT K_SECONDS(4)
-
-/* IRC API */
-#define DEFAULT_SERVER	"irc.freenode.net"
 #define DEFAULT_PORT	6667
 #define DEFAULT_CHANNEL	"#zephyrbot"
+#define MAX_CHANNELS	1
```
Comment (lpereira) [inline samples/net/irc_bot/src/irc-bot.c]: Unused `#define`.

File: samples/net/irc_bot/src/irc-bot.c [hunk 2]
```diff
@@ -43,179 +45,144 @@ static u8_t cmd_buf[CMD_BUFFER_SIZE];
...
+	/* check for a private message from another user */
+	if (!strncmp(nick_buf, cmd, NICK_BUFFER_SIZE)) {
+		on_privmsg_rcvd(umask, umask, space);
+	} else {
+		on_privmsg_rcvd(cmd, umask, space);
 	}
 
-	/* TODO: could be a private message (from another user) */
-	NET_INFO("Unknown privmsg received: %.*s\n", (int)len, cmd);
+	return;
```
Comment (lpereira) [inline samples/net/irc_bot/src/irc-bot.c]: Stray `return;`?

File: samples/net/irc_bot/src/irc-bot.c [hunk 3]
```diff
@@ -240,409 +207,233 @@ process_command(struct zirc *irc, char *cmd, size_t len)
...
-			if (tmp) {
-				tmp = net_frag_read(tmp, pos, &pos, 1, NULL);
-			}
-
+		tmp = net_frag_read(tmp, pos, &pos, 1, cmd_buf + cmd_len);
+		if (*(cmd_buf + cmd_len) == '\r') {
 			cmd_buf[cmd_len] = '\0';
-			process_command(irc, (char *)cmd_buf, cmd_len);
+			process_command((char *)cmd_buf, cmd_len);
+			/* skip the /n char after /r */
```
Comment (lpereira) [inline samples/net/irc_bot/src/irc-bot.c]: Shouldn't this be `\n after \r`?

File: samples/net/irc_bot/src/irc-bot.c [hunk 4]
```diff
@@ -834,239 +632,39 @@ on_msg_rcvd(void *data, struct zirc_chan *chan, char *umask, char *msg)
...
 
-void main(void)
-{
-	struct zirc irc = { };
-	struct zirc_chan chan = { };
-	int ret;
-
-	NET_INFO("Zephyr IRC bot sample");
+	memset(&app_ctx, 0, sizeof(app_ctx));
+	memset(cmd_buf, 0, CMD_BUFFER_SIZE);
```
Comment (lpereira) [inline samples/net/irc_bot/src/irc-bot.c]: These two are zeroed out as part of clearing the BSS, no need to memset() them here.

**Review by lpereira — CHANGES_REQUESTED** [review]
Agreed -- this is a good improvement. Some nits, though.

File: samples/net/irc_bot/src/irc-bot.c (hunk 1 above)
Comment (mike-scott) [reply samples/net/irc_bot/src/irc-bot.c]: Heh, left over from cleanup.  Good catch.

File: samples/net/irc_bot/src/irc-bot.c (hunk 3 above)
Comment (mike-scott) [reply samples/net/irc_bot/src/irc-bot.c]: Fixing.

File: samples/net/irc_bot/src/irc-bot.c (hunk 2 above)
Comment (mike-scott) [reply samples/net/irc_bot/src/irc-bot.c]: Not sure how that got in there.. probably the same.  Thanks!

File: samples/net/irc_bot/src/irc-bot.c (hunk 4 above)
Comment (mike-scott) [reply samples/net/irc_bot/src/irc-bot.c]: Ugh, I had these as parameters and moved them out as statics later.  Another good catch.

**Comment by mike-scott** [comment]
I took some more time and split this up for easy review.   @lpereira sorry about wasting some of your time during the first round.  I addressed all of your comments.

**Review by jukkar — APPROVED** [review]
Looks good, thanks Mike for fixing this!

**Comment by mike-scott** [comment]
@lpereira let me know if you need any other changes.

**Review by nashif — APPROVED** [review]

**Pull request merged**